File: java/core/src/org/daisy/urakawa/property/channel/AudioChannel.java
package org.daisy.urakawa.property.channel;

import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.media.AbstractAudioMedia;
import org.daisy.urakawa.media.IMedia;
import org.daisy.urakawa.media.ISequenceMedia;

/**
 *
 */
public class AudioChannel extends Channel
{
    @Override
    public boolean canAccept(IMedia m) throws MethodParameterIsNullException
    {
        if (!super.canAccept(m))
            return false;
        if (m instanceof AbstractAudioMedia)
            return true;
        if (m instanceof ISequenceMedia)
        {
            for (IMedia sm : ((ISequenceMedia) m).getListOfItems())
            {
                if (!(sm instanceof AbstractAudioMedia))
                    return false;
            }
            return true;
        }
        return false;
    }
}

File: csharp/core/property/channel/AudioChannel.cs
using urakawa.media;
using urakawa.xuk;

namespace urakawa.property.channel
{
    public class AudioXChannel : AudioChannel
    {
        public override string GetTypeNameFormatted()
        {
            return XukStrings.AudioXChannel;
        }
    }

    /// <summary>
    /// A <see cref="Channel"/> that only accepts <see cref="AbstractAudioMedia"/>
    /// </summary>
    public class AudioChannel : Channel
    {

        public override string GetTypeNameFormatted()
        {
            return XukStrings.AudioChannel;
        }

        /// <summary>
        /// Determines if a given <see cref="Media"/> can be accepted by the channel,
        /// which it can if it implements interface <see cref="AbstractAudioMedia"/>
        /// </summary>
        /// <param name="m">The given media</param>
        /// <returns>A <see cref="bool"/> indicating if the given media can be accepted</returns>
        public override bool CanAccept(urakawa.media.Media m)
        {
            if (!base.CanAccept(m)) return false;
            if (m is AbstractAudioMedia) return true;

#if ENABLE_SEQ_MEDIA

            if (m is SequenceMedia)
            {
                foreach (Media sm in ((SequenceMedia)m).ChildMedias.ContentsAs_Enumerable)
                {
                    if (!(sm is AbstractAudioMedia)) return false;
                }
                return true;
            }
#endif //ENABLE_SEQ_MEDIA

            return false;
        }
    }
}